### ISSUE TITLE
Improving the Web Server

### ISSUE THREAD

**Issue by exectails** [aura-project/aura#430]
When we originally added the web server all we wanted was an easy way to add a simple API for the actual server's site and maybe tools to use, and some simple sites, like for guilds, to do basic tasks you can't do in-game.

Over the last months I came to the realization that there's no reason to sell the web server short. It's good enough to run small to medium sized websites on, and even if not for that, it could still be used for prototyping or temporary access to scripts like the registration. The server is there, always running, so why not use it?

However, the current web server has a problem when it comes to extensibility. Not only is it complicated to add new pages, it's even harder to override existing ones, and it's currently impossible to override existing assets, like images and css files.

I'd like to replace our current web server, which uses routing, with one that has a more classic approach. Instead of loading scripts that set up their own routes to make them accessible, you'd simply move the script file to the web folder. Done.

For example, this is part of our registration script.

```
public class RegistrationControllerScript : IScript
{
    public bool Init()
    {
        WebServer.Instance.App.All("/register", (req, res) =>
        {
            // ...
        }
    }
}
```

This goes into a file, you put that file into the script folder, add it to the web script list, and then you can access it via "<url>/register". To change something like the URL, you'd have to modify the script.

What I'm imagining is writing the following, putting it into a file, and simply moving that file to the web folder, wherever I want the file to be.

```
public class RegistrationControllerScript : IScript
{
    public void OnRequest(Request req, Response res)
    {
        // ...
    }
}   
```

As soon as it's in the folder, you can access it via "<url>/register.cs" for example, or whatever _you_ called the file. It's a simple file serving web server, with support for C# scripting. Every file _you_ move to the web folder becomes accessible. All integrated with the user folder of course, that's treated with higher priority, so you can replace system scripts, images, templates, CSS files, etc.

With this setup we could also easily support other scripting languages if we wanted to, like PHP. It could be possible to run any PHP software, like forums, from our server.

Sounds like I'm trying to sell something innovative here, when it's really the old-school way of handling web requests =P I've made a [prototype](https://github.com/exectails/Swebs) a while ago, and it really wouldn't be all that hard to switch them out.

Theoretically the user system could work with the current web server as well, overriding pages, scripts, and files could be made possible, but I'm thinking the current web server is not exactly user friendly. You can't even put a simple HTML file on it without knowing exactly where to put it, and even if that's made easier we would have 2 separate systems at work, static file serving and the script list.

Not to mention, what if you download something like a CMS that is supposed to run on Aura's web server? It needs to go into the right folder, the scripts need to be added to the list, there could be conflicts in the paths, because the scripts decide them, it's problematic in my eyes.

Tell me what you think!

**Comment by DennouNeko** [comment]
Being able to run everything from a single server would make things a lot easier. In current setup one would have to set up Apache, nginx or some other web server to handle php scripts and add rewrite rules to pass certain requests to Aura.
If Aura's web server was able to handle php and maybe cgi-bin like requests it would allow for a lot better integration of services with much easier configuration.

As for putting scripts in web servers tree, if any extension could be used we'd have to have a way to identify them (shebang?), or it could be required for them to have .cs extension with some support for .htaccess, rewrite rules and index search order (like index.cs, index.php, index.html by default).

**Comment by exectails** [comment]
Yea, we would need common web server features that we don't really have atm. For defining how certain file types are handled, Swebs has the [Configuration](https://github.com/exectails/Swebs/blob/master/Swebs/Configuration.cs#L91), same for [index order](https://github.com/exectails/Swebs/blob/master/Swebs/Configuration.cs#L69). A mod rewrite like system would be pretty easy as well, seeing how it's basically just regex.

**Comment by majorcyto** [comment]
Improvement's would be welcome for more customizing. As said above PHP , or CGI-Bin requests would be interesting. As i think you mentioned before Exec a simple message board would be a neat addition to guild tools/features.

It could just be a "list" of topics, that splits into pages. With the ability to pin a topic at the top. Each guild could automatically have one, or buy one with gold (yay another guild gold sink. :P). Officers are automatically mods of it, boom.

Another interesting feature, would be an actual Guild Store of some kind built in as well (separate from the normal "shop"). Accessed through the guild menu, you could buy things with guild gold. Say a guild wide buff for 1 hour, or 10 hours, etc. Possibilities are endless but i'm just offering ideas on what it could be used for.

But yeah, the current one is a bit hard to customize compared to how it could be set up i guess.

**Comment by exectails** [comment]
I've switched from SharpExpress to Swebs in a branch, it would be good if those interested in the web server could take a look at it, see how you like it.

https://github.com/aura-project/aura/tree/add-swebs
https://github.com/aura-project/aura/tree/add-swebs/system/web
https://github.com/aura-project/aura/blob/add-swebs/src/WebServer/WebServer.cs#L96

As you can see in the system's web directory, all web files are now in this one folder, with the web server acting like a traditional web server, all files in there being accessible. And as described above, Swebs first looks at the user's web folder for files, so you can easily replace images, styles, etc. just by putting them there.

New pages can be added the same way, by simply placing them in one of the web folders.

There's no access rules or rewriting yet, but I think those are secondary.

**Comment by QewQew** [comment]
Is there more information about swebs (e.g. Github repo, home page) that we can take a look at?

**Comment by exectails** [comment]
@QewQew I had linked to it in-line above, but here's the direct link: https://github.com/exectails/Swebs

**Comment by exectails** [comment]
Closing, since the branch was merged in 6f9a39625e2e405fdcac6a916ad11bc86e2916c0.
